### PR TITLE
Fullscreen Mode: Change the "Back" button to toggle the sidebar

### PR DESCRIPTION
**Pull request by shaunandrews**
With the recent move to make Fullscreen mode the default, I've heard from many people that they expect to be able to toggle the WordPress sidebar menu by clicking on the WordPress logo in the top left. This PR tries that:

Fullscreen Sidebar Toggle i1

### CODE REVIEW

**Comment by MichaelArestad** [comment]
I want this and I want it now! <3

**Comment by mapk** [comment]
Moving this functionality to the top level is great. While it doesn't provide the top bar, it's still usable and surfaces the wp-admin navigation with ONE click. Great exploration!

**Comment by SchneiderSam** [comment]
This is soooooo cool.

**Comment by paaljoachim** [comment]
This is awesome! It sure makes going into Full Screen mode a lot easier.

Along with this PR let's also get a shortcut in place. https://github.com/WordPress/gutenberg/issues/20626

**Comment by shaunandrews** [comment]
A couple of things I could use help with:

- I think when you toggle the sidebar open, the keyboard `:focus` should jump to the wp-admin sidebar.
- Right now, the sidebar sticks around until you click the WP logo again. Part of me thinks that clicking anywhere _outside_ of the sidebar should hide the sidebar again. In fullscreen-mode, I don't think the sidebar should stick around.

I'm not sure how to go about implementing either of the above, but would love to learn. All tips, tricks, and commits welcomed.

**Comment by paaljoachim** [comment]
Testing out icons for bringing back the sidebar menu.

Original.
Sidebar-W-original

Grey W.
Sidebar-W-Gutenberg-5

White W.
Sidebar-W-Gutenberg-7

Grey W with vertical line.
Sidebar-grey-W-with-line

Grey W with dashes and vertical line.
Sidebar-W-Gutenberg-1

White W with dashes and vertical line.
W-show-sidebar-Gutenberg-2

White W with dashes and vertical line 2.
Show-sidebar-W-Gutenberg-3

Grey W with horizontal dashes.
Sidebar-W-Gutenberg-4

**Comment by bph** [comment]
This is an awesome idea, and well executed. Thank you @shaunandrews :-) Sorry, I can't help you with development but I wanted you to know that a lot of end users will be quite happy with this.

**Comment by mtias** [comment]
Thanks for exploring this @shaunandrews. We should look at the centered logo a bit in the open state: it'd make sense to see if we can render the site title when it expands (and keep it hidden when collapsed) to balance it better.

**Comment by ItsJonQ** [comment]
> A couple of things I could use help with:
>
> * I think when you toggle the sidebar open, the keyboard :focus should jump to the wp-admin sidebar.
>
> * Right now, the sidebar sticks around until you click the WP logo again. Part of me thinks that clicking anywhere outside of the sidebar should hide the sidebar again. In fullscreen-mode, I don't think the sidebar should stick around.

@shaunandrews Haii!! Great work on this update! I can certainly help you with these items :)

re: Focus jump:

To do this, we'll need to hook into some sort of `onOpen` state transition to make the focus. I think maybe keeping the focus on the "W" logo may be better. However, perhaps when...

* Focus is on "W" logo, and...
* Menu is open, and...
* User presses `DOWN` on the keyboard,...
* `:focus` jumps to the first menu item

re: Closing menu when clicking outside

This is often accomplished using 2 techniques:

1. "Overlay"

When the menu is open, an "Overlay" (a transparent `div`) is rendered on top of the Editor, example:

overlay

This *blocks* interaction for anything underneath it, and allows you to easily attach an `onClick` handler on it. Which means, when you click on the overlay, the menu can close.

2. "Clicking outside"

This is the trickier, but often more preferrable, way to handle these interactions. What needs to be done is a `click` event listener needs to be added, and logic needs to read something like:

* If the (clicked) `event.target` is...
* **NOT** the "W" logo, and...
* **NOT** the menu, and...
* **NOT** the top toolbar,....
* Then close.

Hope this helps 🙏

**Comment by mapk** [comment]
This is beautiful! 
I'm in favor of seeing a prototype that works in Matias' suggestion. Keep the logo fixed to the left and add the site title when expanded. 

This is also the desire after [today's triage meeting in slack](https://wordpress.slack.com/archives/C02S78ZAL/p1585670695090100).

**Comment by ItsJonQ** [comment]
Haiii!! I coordinated with @shaunandrews .
I'm gonna try to help out a bit with the JS code side.

🤞

**Comment by ItsJonQ** [comment]
Heya @shaunandrews !

I have updates :). I've made some refactors and enhancements in a separate branch:
https://github.com/WordPress/gutenberg/tree/try/fullscreen-sidebar-toggle-refactor

(This branch was also rebased with latest `master`)

You can see my changes here:
https://github.com/WordPress/gutenberg/commit/9a9f9efeffdaeb2d71d0a00d4798d63a67a9f44b

TLDR:
* Create a custom React hook to better manage the open/close menu state
* Leveraged hook to bind the click + keyboard events for closing or (re)focusing

Let me know what you think!

---

Note: I haven't done anything with Site Title yet.

> it'd make sense to see if we can render the site title when it expands (and keep it hidden when collapsed)

We can do this. However, I'm not too sure what it may look like if the site title was **longer**. In the regular WP Admin, the "W" logo and title are separate from the menu, allowing longer titles to flow.

I think a mockup of this would be helpful 😊

Thank you!

**Comment by shaunandrews** [comment]
> Let me know what you think!

Super cool, @ItsJonQ — thanks for your help on this. I'd love to see your branch merged into this one.

One thing I'm still unsure about is the keyboard interaction. Having to press the down arrow to start navigating the wp-admin sidebar seems strange. If you look at how the inserter works, when you tab to it and select, the `:focus` goes into the popover. I feel like this maps to the wp-admin toggle; Selecting the logo with the keyboard seems like it should move the `:focus` into the menu.

--

As for the site title, I like the idea in theory. But I'm not sure there's enough room for it with the current width of the menu. Here's what it would look like with a truncated site title:

image

We _could_ change the width of the menu (only within the Editor context) to add some overall space, but I'd expect that to feel strange in use:

image

We could break the title down to its own line:

image

**Comment by ItsJonQ** [comment]
@shaunandrews Thanks for 🍐 'ing with me on this one!

I just rebased my changes from my `refactor` branch onto yours and pushed it up.

One thing we noticed was that the name of `FullscreenModeClose` no longer described what the component did.

I renamed it to `AdminMenuToggle`, and have adjusted CSS classNames and E2E tests accordingly.

**Comment by ItsJonQ** [comment]
> Selecting the logo with the keyboard seems like it should move the :focus into the menu.

@shaunandrews Ah yes, the keyboard interactions are tricky for this one. I just pushed an update that focuses using `TAB` rather than `DOWN`.

Let me know how this feels :)

**Comment by enriquesanchez** [comment]
Keyboard interaction feels good to me, but I found one small issue.

If I move focus to the "W" and hit Enter, the wp-admin menu expands and my focus stays on the "W". This is expected. I can then move focus to the first item in the menu by hitting `TAB`.

The one problem I found was going back. If try to do `shift + TAB` to go back up to the "W" button, I can't seem to be able to land focus on it. I do see a "Skip to toolbar" skip link, which then takes me to the "W" button if I do one more `TAB`, but that was unexpected. I expected to be able to go back the same way I came down.

If I use the "Skip to toolbar" link, I'm not really taken to the "toolbar" because if I do a `TAB` again after focusing on the "W", my focus moves down to the "Dashboard" in the menu. That is unexpected, I assumed that because I'm now on the toolbar I would move focus to the inserter button instead.

I would need to collapse the menu first, so that I'm now on the toolbar and can move to the inserter.

**Comment by shaunandrews** [comment]
This breaks the top toolbar:

image

**Comment by timhibberd** [comment]
@shaunandrews I like this a lot but this would tie a navigation function to an icon that a lot of MSPs hide for business sites. So the question is...if this is hidden can we be assured of no side-effects? Will the icon that inherits that position now inherit the navigation logic?

If you are going to do this, anchor it as a mandatory navigation element with a php API to replace it with a different icon / function. Even that would break on legacy sites who don't know about the new API.

**Comment by noahtallen** [comment]
@timhibberd It looks like this is planned as part of https://github.com/WordPress/gutenberg/issues/20929

**Comment by timhibberd** [comment]
Excellent. Love it (combo of this PR + #20929 )  :-)

**Comment by Clorith** [comment]
My initial thought is to not move the WP logo, this is supposed to bridge the gap between "losing" the sidebar, and it should retain the same visual appearance of the sidebar then.

File: packages/block-editor/src/components/fullscreen-mode/style.scss [hunk 1]
```diff
@@ -23,7 +22,6 @@ body.js.is-fullscreen-mode {
 			transition: left 0.12s ease-in-out;
 		}
 
-		// This is all so stupid
```
Comment (noahtallen) [inline packages/block-editor/src/components/fullscreen-mode/style.scss]: 🤣

Comment (shaunandrews) [reply packages/block-editor/src/components/fullscreen-mode/style.scss]: lol, I often make a lot of comments as I'm working on code. Most get deleted. They usually include much more... harsh... language. 😆

Comment (ItsJonQ) [reply packages/block-editor/src/components/fullscreen-mode/style.scss]: 😂

File: packages/block-editor/src/components/fullscreen-mode/style.scss [hunk 2]
```diff
@@ -10,27 +9,69 @@ body.js.is-fullscreen-mode {
 			height: calc(100% + #{ $admin-bar-height });
 		}
 
-		#adminmenumain,
+		// Hide the top adminbar toolbar.
```
Comment (tellthemachines) [inline packages/block-editor/src/components/fullscreen-mode/style.scss:12]: We might need to add and remove the `display: none` with JS (as adding it directly in the CSS might ruin the animation). The problem is that currently all the nav elements, and the skip links, are still focusable when hidden, and that causes some weirdness when tabbing through them:

Screen Shot 2020-04-03 at 4 25 08 pm

Screen Shot 2020-04-03 at 4 24 54 pm

**Review by tellthemachines — COMMENTED** [review]
Looks great overall, just one small issue below 🙂

File: packages/block-editor/src/components/fullscreen-mode/style.scss (hunk 2 above)
Comment (shaunandrews) [reply packages/block-editor/src/components/fullscreen-mode/style.scss:12]: Ah, good catch. I _think_ adding `pointer-events: none;` might work here. Using `display: none;`, like you said, will break the animation. I'll see what I can do to resolve this — thanks for taking a look.

Comment (ItsJonQ) [reply packages/block-editor/src/components/fullscreen-mode/style.scss:12]: @tellthemachines + @shaunandrews Haiii! I just pushed something up that resolves this (I believe).

It hides the element wrapping the menu items. This allows for the outer (bar) to retain it's animation transition properties, but also prevents the inner links from being tabbable.

File: packages/interface/src/components/fullscreen-mode/style.scss [hunk 3]
```diff
@@ -10,14 +9,80 @@ body.js.is-fullscreen-mode {
...
+		left: 0;
+	}
+
+	// Show the admin menu
+	#adminmenuwrap {
+		display: block;
+	}
+
+	.edit-post-header-toolbar__block-toolbar,
+	.block-editor-editor-skeleton__body,
```
Comment (vindl) [inline packages/interface/src/components/fullscreen-mode/style.scss:64]: I renamed this to `InterfaceSkeleton` in https://github.com/WordPress/gutenberg/pull/21335, so the classnames here will need to be updated.

File: packages/edit-post/src/components/header/admin-menu-toggle/index.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+ * WordPress dependencies
+ */
+import { useSelect } from '@wordpress/data';
+import { Button } from '@wordpress/components';
+import { useRef, useEffect, useState } from '@wordpress/element';
+import { ESCAPE, TAB } from '@wordpress/keycodes';
+import { __ } from '@wordpress/i18n';
+import { wordpress } from '@wordpress/icons';
+
+function AdminMenuToggle() {
```
Comment (vindl) [inline packages/edit-post/src/components/header/admin-menu-toggle/index.js:11]: Since we'll need the same component in Site Editor and maybe Widgets screen, could we move it to `@wordpress/interface` packaged so it can be more easily reused in that context?

**Comment by paaljoachim** [comment]
Which shortcut should we use for going in and out of Fullscreen mode?
I am asking as I have an issue in relation to it: https://github.com/WordPress/gutenberg/issues/20626

File: packages/interface/src/components/fullscreen-mode/style.scss (hunk 3 above)
Comment (shaunandrews) [reply packages/interface/src/components/fullscreen-mode/style.scss]: Thanks for the heads up. I've updated the classnames.

File: packages/edit-post/src/components/header/admin-menu-toggle/index.js (hunk 4 above)
Comment (shaunandrews) [reply packages/edit-post/src/components/header/admin-menu-toggle/index.js]: Cool, I managed to figure out how to move it to the interface package. Should be good to go now.

File: packages/edit-post/src/components/header/index.js [hunk 5]
```diff
@@ -6,12 +6,12 @@ import { Button } from '@wordpress/components';
...
 import { useSelect, useDispatch } from '@wordpress/data';
 import { cog } from '@wordpress/icons';
-import { PinnedItems } from '@wordpress/interface';
+import { PinnedItems, AdminMenuToggle } from '@wordpress/interface';
 
 /**
  * Internal dependencies
  */
-import FullscreenModeClose from './fullscreen-mode-close';
+//import AdminMenuToggle from './admin-menu-toggle';
```
Comment (youknowriad) [inline packages/edit-post/src/components/header/index.js:14]: I guess this should be removed

File: packages/interface/src/components/admin-menu-toggle/index.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+import { __ } from '@wordpress/i18n';
+import { wordpress } from '@wordpress/icons';
+
+function AdminMenuToggle() {
+	const buttonRef = useRef();
+	const toggleMenu = useToggle( { ref: buttonRef } );
+
+	const { isActive, postType } = useSelect( ( select ) => {
+		const { getCurrentPostType } = select( 'core/editor' );
+		const { isFeatureActive } = select( 'core/edit-post' );
```
Comment (youknowriad) [inline packages/interface/src/components/admin-menu-toggle/index.js:17]: This component has been moved to a generic package "interface" which means it can't use a specific store like "core/edit-post" or "core/editor". I'm not even certain why we're checking here the post type here. For me we should remove all that use select code and the check to return null a bit under.

We should just not render this component from its parent when we're not on fullscreen.

File: packages/interface/src/components/admin-menu-toggle/index.js [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+		};
+	}, [] );
+
+	if ( ! isActive || ! postType ) {
+		return null;
+	}
+
+	return (
+		<Button
+			className="edit-post-fullscreen-admin-menu-toggle"
```
Comment (youknowriad) [inline packages/interface/src/components/admin-menu-toggle/index.js:32]: This classname should probably be updated to reflect the new package.

File: packages/interface/src/components/admin-menu-toggle/style.scss [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+		&:focus {
+			box-shadow: none;
+
+			svg {
+				box-shadow: 0 0 0 1.5px color($theme-color);
+			}
+		}
+
+		body.js.is-showing-admin-menu & {
```
Comment (youknowriad) [inline packages/interface/src/components/admin-menu-toggle/style.scss:45]: Why do we need the `js` className on all these selectors?

File: packages/interface/src/components/fullscreen-mode/style.scss [hunk 9]
```diff
@@ -10,14 +9,80 @@ body.js.is-fullscreen-mode {
...
+		#adminmenu {
+			position: relative;
+			bottom: auto;
+		}
+
+		#adminmenuback {
+			top: 60px;
+		}
+
+		.edit-post-header-toolbar__block-toolbar,
```
Comment (youknowriad) [inline packages/interface/src/components/fullscreen-mode/style.scss:40]: We shouldn't be targeting edit-post styles on the interface package. What is this about?

Comment (youknowriad) [reply packages/interface/src/components/fullscreen-mode/style.scss:40]: Looks like there's more of these styles a bit under. can't we target the skeleton header instead?

File: packages/edit-post/src/components/header/index.js [hunk 10]
```diff
@@ -58,7 +58,7 @@ function Header() {
 
 	return (
 		<div className="edit-post-header">
-			<FullscreenModeClose />
+			<AdminMenuToggle />
```
Comment (youknowriad) [inline packages/edit-post/src/components/header/index.js:61]: Can we do the same for the edit-site package (Site builder)? This will help you figure out better how to make these styles generic and work across screens.

**Review by youknowriad — COMMENTED** [review]
Great work here.

File: packages/edit-post/src/components/header/index.js (hunk 5 above)
Comment (shaunandrews) [reply packages/edit-post/src/components/header/index.js]: Yup. Gone.

File: packages/interface/src/components/admin-menu-toggle/index.js (hunk 7 above)
Comment (shaunandrews) [reply packages/interface/src/components/admin-menu-toggle/index.js]: Nice catch. Updated.

File: packages/interface/src/components/admin-menu-toggle/style.scss (hunk 8 above)
Comment (shaunandrews) [reply packages/interface/src/components/admin-menu-toggle/style.scss]: I guess we don't? I'm not sure why I added it.

File: packages/interface/src/components/fullscreen-mode/style.scss (hunk 9 above)
Comment (shaunandrews) [reply packages/interface/src/components/fullscreen-mode/style.scss]: I see. That makes sense. I've updated it.

**Comment by mapk** [comment]
It's going to be interesting to see how this works in relation to https://github.com/WordPress/gutenberg/pull/20951. Any thoughts around how these two slide-outs work together, @shaunandrews?

File: packages/interface/src/components/admin-menu-toggle/index.js (hunk 6 above)
Comment (shaunandrews) [reply packages/interface/src/components/admin-menu-toggle/index.js]: Took me a bit to understand this, but now it makes sense. I've removed the display logic out of the component and instead moved it into where it's used.

**Comment by shaunandrews** [comment]
> Any thoughts around how these two slide-outs work together

The wp-admin menu will simply push the inserter over, just like the rest of the editor's skeleton. I don't foresee any interference or issues.

**Comment by ItsJonQ** [comment]
Looks like Travis is unhappy. Let me see if I can help! 💪

**Comment by shaunandrews** [comment]
> Any thoughts around how these two slide-outs work together

The wp-admin menu will simply push the inserter over, just like the rest of the editor's skeleton. I don't foresee any interference or issues.

File: packages/edit-post/src/components/header/index.js [hunk 11]
```diff
@@ -6,12 +6,11 @@ import { Button } from '@wordpress/components';
 import { PostPreviewButton, PostSavedState } from '@wordpress/editor';
 import { useSelect, useDispatch } from '@wordpress/data';
 import { cog } from '@wordpress/icons';
-import { PinnedItems } from '@wordpress/interface';
+import { PinnedItems, AdminMenuToggle } from '@wordpress/interface';
```
Comment (joyously) [inline packages/edit-post/src/components/header/index.js:9]: I know it's not a component, but the _is_ an existing thing that collapses the admin menu. I'm not sure what it's called, but it's at the bottom of the menu. I think using the name `AdminMenuToggle` could be a bit confusing since one already exists. I'm not sure why the existing one isn't used for this fullscreen thing.

Comment (shaunandrews) [reply packages/edit-post/src/components/header/index.js:9]: I guess I think of the `#collapse-button` option that exists at the bottom of the menu as a different feature; The collapse button changes the width of the admin sidebar, not the visibility.

This new `AdminMenuToggle` component toggles the visibility.

The two features can work together; You can collapse the admin sidebar _and_ toggle its visibility.

File: packages/interface/src/components/admin-menu-toggle/index.js [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+			if ( keyCode === ESCAPE ) {
+				closeAdminMenu();
+			}
+
+			if ( keyCode === TAB ) {
+				focusFirstAdminMenuItem();
+			}
+		};
+
+		if ( isActive ) {
```
Comment (gziolo) [inline packages/interface/src/components/admin-menu-toggle/index.js:98]: How about you move the check for `isActive` to `handleOnKeyDown` callback to make the setup of event handlers independent from React state changes?

File: packages/interface/src/components/fullscreen-mode/style.scss [hunk 13]
```diff
@@ -10,14 +9,65 @@ body.js.is-fullscreen-mode {
...
+		}
+
+		#adminmenuback {
+			top: 60px;
+		}
+
+		.interface-interface-skeleton__body,
+		.interface-interface-skeleton__footer {
+			transition: margin-left 0.12s ease-in-out;
+		}
```
Comment (youknowriad) [inline packages/interface/src/components/fullscreen-mode/style.scss:43]: All these `skeleton-*` styles on this stylesheet make the "full screen mode" work properly if your screen is using the Skeleton. do you think it's possible to make it independent of that? I mean no matter how the current screen is rendered, it could work?

I don't see this as a blocker but more "good to have"

**Review by youknowriad — APPROVED** [review]
on Safari, the wp menu doesn't cover the full height of the window.  (see the white space at the bottom)

Capture d’écran 2020-04-15 à 9 49 13 AM

Other than that, I think this is looking good.

File: packages/interface/src/components/admin-menu-toggle/index.js (hunk 12 above)
Comment (ItsJonQ) [reply packages/interface/src/components/admin-menu-toggle/index.js:98]: Oo! That works for me :) Thank you 🙏

**Comment by shaunandrews** [comment]
Rebased and fixed the display bug in Safari.

**Comment by jasmussen** [comment]
The left menu doesn't scroll when in fullscreen mode on a small viewport:

Screenshot 2020-04-24 at 14 58 32

The "fix" in non-fullscreen mode is to allow the HTML element to scroll, causing a double scrollbar on the right. Not an ideal fix, and would love ideas for making that better.

**Comment by mtias** [comment]
I think we need to revert this one and continue working on it for a bit. There are aspects that need to be better handled for an adequate integration, otherwise states like this are possible:

image

**Comment by aduth** [comment]
Revert pull request at #21929
